cache/cache.go

```diff
@@ -47,42 +49,48 @@ type randomEvictedCache struct {
 func NewRandomEvictedCache(maxBytes int) Cache {
 	return &randomEvictedCache{
 		maxBytes: maxBytes,
-		data:     make(map[string]memorizedMeasurable),
+		data:     make(map[string]memoizedMeasurable),
 	}
 }
 
+func (c *randomEvictedCache) evictOneLocked() {
+	i := int(rand.Int63()) % len(c.keys)
+	last := len(c.keys) - 1
+	var toRemove string
+	toRemove, c.keys[i] = c.keys[i], c.keys[last]
+	c.cachedBytes -= 2*len(toRemove) + c.data[toRemove].Size()
+	delete(c.data, toRemove)
+	c.keys = c.keys[:last]
+}
+
 // Get impelments the Cache interface.
 func (c *randomEvictedCache) Get(key string) (data Measurable, ok bool) {
 	c.mu.RLock()
 	defer c.mu.RUnlock()
-	memorized, ok := c.data[key]
+	memoized, ok := c.data[key]
 	if !ok {
 		return nil, false
 	}
-	return memorized.m, ok
+	return memoized.m, ok
 }
 
 // Add implements the Cache interface.
 func (c *randomEvictedCache) Add(key string, data Measurable) {
-	memorized := memorizedMeasurable{m: data}
-	if len(key)+memorized.Size() > c.maxBytes {
+	memoized := memoizedMeasurable{m: data}
+	increase := 2*len(key) + memoized.Size()
+	if increase > c.maxBytes {
 		return
 	}
 	c.mu.Lock()
 	defer c.mu.Unlock()
 	if _, ok := c.data[key]; !ok {
-		c.cachedBytes += len(key) + memorized.Size()
+		c.cachedBytes += increase
 	}
-	if c.cachedBytes > c.maxBytes {
-		for k, v := range c.data {
-			delete(c.data, key)
-			c.cachedBytes -= len(k) + v.Size()
-			if c.cachedBytes <= c.maxBytes {
-				break
-			}
-		}
+	for c.cachedBytes > c.maxBytes {
+		c.evictOneLocked()
```

Comment:
FWIW this isn't a uniform distribution: the items that are left after each removal get disproportionate consideration for removal. It should be fine in the long run though.